### ISSUE TITLE
license changing

### ISSUE THREAD

**Issue by take-cheeze** [EasyRPG/Player#167]
Since all the code copyright holder agreed to move to non-GPL new license we need to prepare for it.
The issues for it will be
1. which license to move on
-> strong candidate would be MIT license.
2. when to move on
-> I'm thinking of moving to new license when the mruby branch is merged.
3. how do we treat binary distribution linking GPL library
-> maybe we need to remove those libraries or stop distributing it.
4. license text change of source file
-> we'll use some text replace program like sed

----

- [x] 20kdc
- [x] Albeleon
- [x] BlisterB
- [x] carstene1ns
- [x] ChristianBreitweiser
- [x] Desdaemon
- [x] drxgb
- [x] ell1e
- [x] eneway (Narcodis)
- [x] fdelapena
- [x] florianessl
- [x] fmatthew (mateofio)
- [x] Ghabry
- [x] glynnc
- [x] jetrotal
- [x] johannschopplich
- [x] JorgeMaker8000
- [x] kakurasan
- [x] MackValentine
- [x] MakoInfused
- [x] MarianoGNU
- [x] PrimeKick
- [x] Rikku
- [x] Rinnegatamante
- [x] rohkea
- [x] Rueta
- [x] scurest
- [x] sorlok
- [x] take-cheeze
- [x] Toolman2k
- [x] Tondorian
- [x] tyrone-sudeium
- [x] vgvgf
- [x] Zegeri
- [x] Zhek

----

If you cannot find your name in the list above:

- This happens when your change was _trivial_, this applies to stuff like fixing of typos (or other single character changes), only deleting code and reordering of existing code (without inventing new code). 
- You are also not asked when the contribution was WQY font related (this part stays GPL due to the font license).
- You are also not asked when your change affects only homebrew platforms and other exotic platforms (often dependencies used here are under GPL so this would change nothing)
- You contributed only to the Android User Interface: This is already under MIT
- You contributed only (Android) translations: They are already under CC0

When you think that you made a significant contribution and cannot find your name in the list above: We are sorry. Please contact us.

**Comment by fdelapena** [comment]
This issue is quite outdated, the current situation is:

- Already switched to SDL2 from SDL1.2.
- Already switched to ICU from iconv.
- Already switched to mpg123 from libmad. There are no GPL deps anymore.
- App Store allows LGPL dynamic linking, as copyright holders of the package there is no problem with license issues. And yes, [iOS has a way to use dynamic libraries](http://stackoverflow.com/a/10509453/1244033) and even since iOS 8 officially supports dylibs, so policies have been relaxed.
- 2k/3 English release, still "supported".
- According to the vendor, 2k/3 original source code is lost.

Because we don't have limitations with keeping Player as GPL and not a good reason to switch to MIT nowadays, and because the 2k/3 "code lost" issue might be interested on a non-copyleft version for a takeover (liblcf is MIT licensed already for their needs), what do you think about keeping Player GPL and closing this one as `wontfix`?

**Comment by Ghabry** [comment]
You make valid points but there is one thing that bugs me:
I would like to see some kind of "linking exception" in the license which allows us to interface with non-system libs. As an example I would name the steamworks library to provide steam features like achievements (other stores (GOG?) probably have similiar libs). There are probably other cases where it makes sense, but these are the first ones I thought of.
The library is non-free and not a system library, we are not allowed to link with it and call into the functions :(. But the lib doesn't interact with our code at all, so I'm not seeing a problem there, it behaves like a system lib.
tl;dr: An exception to allow linking and interacting (Player calls into Lib) with non-free libraries if they don't depend on Player. The GPL applies for the implementation of the interface obviously because it modifies Player code.

**Comment by fdelapena** [comment]
Yeah, I've totally forgot this point. Indeed an exception is really convenient for these cases.

By the way, any good related real world GPL exception from existing projects to adopt?
If possible, some variant which is already OSI/DFSG/FSF approved to help updating package maintainers without major bureaucratic issues.

**Comment by Ghabry** [comment]
Yeah imo this is not a problem for distribution because we will not link against these non-free APIs in the normal case. This would be just for special needs, like game developers wanting to use these APIs.

Is maybe a good idea to put the whole non-free API in seperate files to make distribution easier for the OSI conform tarballs as you suggested. And all non-GPL files should be manual opt-in via a configure argument.

The suggested exception on the GNU page is a bit specific, we should try to generalize it. Or we just get the permission by all devs that adding additional libraries on demand is allowed and then we just extend the license everytime ^^

```
Additional permission under GNU GPL version 3 section 7

If you modify this Program, or any covered work, by linking or combining it with any library listed in GPL_EXCEPTION.txt (or a modified version of these libraries), containing parts covered by the terms of the respective license of these libraries as listed in GPL_EXCEPTION.txt, the licensors of this Program grant you additional permission to convey the resulting work.
```

**Comment by carstene1ns** [comment]
My main concern is about loosing statically linked ports.
Usually most homebrew libraries are GPL, changing our license will violate their license.

While using dynamic linking the (open) license does not matter at all, when we not change the used library.

**Comment by Ghabry** [comment]
After a bit of research I think we have to scratch this approach for now. Valve also requires you to sign a NDA for the full steamworks API and you are not allowed to publish any code that is calling into there API. Ancurio wrote a story on Reddit about publishing To The Moon with Steam achivements...
But when the editor is ready (in 2 years? :D) we really need a solution for this because it limits the ways how game developers can sell there games :/

So to do the exception part it would get even more chaotic: Write a propreitary library that links against Steam, mention this library in the exception list, link against this library --> Keeps the Steam code secret m(

About your text: Yeah, then it should be dual licensed... one GPL and one GPL with that exception. The exception would only apply for that corner case, all linux distributions and homebrews could use the normal GPL version because they don't need the non-free part.

**Comment by carstene1ns** [comment]
About game devs selling games including player:
What matters is their content, they do not sell Player. IMO also all published games should need to have a `README/LICENSE-EasyRPG-Player` file describing what it is.

About steam api integration:
What about doing it the other way around? Add a linking exception to Player, that allows including in closed source (only unmodified, or with changes published of course).
Then let them write a puppeteer executable that links `libeasyrpg-player`. :grinning:

**Comment by Ghabry** [comment]
This sounds like an option. Allows exactly what I wanted anyway with that library exception: Changes to Player count as derivative work: Must be published under GPL and the other direction doesn't matter and is allowed propreitary.

But there is another issue: Propreitary (and NDA) for video game console SDKs. They need a way to handle Video/Audio/Input of the Player. Therefore I would like to see another exempt for classes that implement the BaseUi interface.

These exceptions should be enough to make the Player usable in 99% of all cases I can think of :)

**Comment by Ghabry** [comment]
Is the "you must ship a license file" part of the GPL?

```
Additional permission under GNU GPL version 3 section 7

Linking this library statically or dynamically with other modules is making a combined work based on
this library. Thus, the terms and conditions of the GNU General Public License cover the whole 
combination.
As a special exception, the copyright holders of this library give you permission to link this library with 
independent modules to produce an executable, regardless of the license terms of these independent
modules, and to copy and distribute the resulting executable under terms of your choice, provided that
you also meet, for each linked independent module, the terms and conditions of the license of that 
module. An independent module is a module which is not derived from or based on this library.

As special exceptions, the copyright holders of this library give you permission to:
a)
link this library with independent modules to produce an executable, regardless of the license terms
of these independent modules, and to copy and distribute the resulting executable under terms of
your choice, provided that you also meet, for each linked independent module, the terms and 
conditions of the license of that module. An independent module is a module which is not derived
from or based on this library.

b)
link this library with modules that explicitly derive from unmodified versions of the classes listed in GPL_EXCEPTIONS.TXT and to copy and distribute the derived classes under terms of your choice.

If you modify this library, you may extend this exception to your version of the library, but you are not obliged to do so. If you do not wish to do so, delete this exception statement from your version.
```

**Comment by Ghabry** [comment]
Here is a cool list of license exceptions: https://spdx.org/licenses/exceptions-index.html

Especially useful is the Qwt and U-Boot exception because it contains stuff I want :D

So I would change my ```b)``` to:

```
b) The classes BaseUi (base_ui.h) and AudioInterface (audio.h) and the namespaces FileFinder (filefinder.h) and Main_Data (main_data.h) define interfaces required for system interoperability. Classes derived from unmodified versions of these classes and implementations of unmodified versions of these namespaces are not considered a derived work.
```

Things that are not stated as exception but are obviously not a derived work in my opinion: System initialisation code required before invoking Player::Init and including header files if that include has no side effects on the Player.

**Comment by fdelapena** [comment]
:+1:, however in case of particular API changes there (namespaces <-> classes) might need to update the license exception later.

**Comment by Ghabry** [comment]
Just to get this finished directly after 0.5 release.
As stated in the main post end of 2013 all devs agreed on relicensing to another license. Back at this time I contacted the devs that were not contributing anymore, I even found the e-mail...

In the meanwhile a few further devs joined, git blaming says:
@Zegeri (tons of interpreter fixes)
@scurest (interpreter fixes)
@ChristianBreitwieser (speexdsp and libretro port)
@Tondorian (minor battle system fixes)
@Rinnegatamante Wrote PSVita and 3DS port, doesnt need relicensing but would be nice for convenience.

You five: _Would you agree on adding a linking exception to our GPL license to allow interoperability with propreitary code (and NDA) like the Steamworks Library (Achievements & co.) to make the Player more useful for game developers?_


The developers that agreed to relicensing in general were:
MarianoGNU
glynnc
Rikku
take-cheeze
vgvgf
Zhek
fdelapena
Ghabry

I assume carstene1ns already agreed.

People that don't need asking:

Rohkea contributed fonts and his work was under public domain.
BlisterBoy codes for the Game browser only which is already MIT (at least I used this as the license for my game browser, not sure if the header is still there >.>) and is not a derived work of Player.

**Comment by scurest** [comment]
> Would you agree...

Sure.

**Comment by Rinnegatamante** [comment]
Me too i agree, no problem.

**Comment by Zegeri** [comment]
I agree.

**Comment by ChristianBreitwieser** [comment]
I agree.

**Comment by carstene1ns** [comment]
Just leaving this here regarding steam: https://hg.icculus.org/icculus/steamshim/file/tip/README.txt
> You have a GPL-licensed game and can't link directly against the Steamworks
> SDK. The child process links against the simple, open source C code, which
> talks to the open source C++ code via a pipe, which talks to Steamworks. You
> can now add Steam achievements to your game without violating the GPL.

**Comment by Ghabry** [comment]
To simplify this (and carstene1ns already suggested something like this) a "core" and a "player" (find a better name) library could be created.
The "Core" gets a lax license and the "Player" lib gets a linking exception.

The requirement for "Core" is that the "Player" lib depends on it but the "Core" may not depend on "Player".

Therefore core should contain (havn't verified if it still compiles with only these files, will update when I find more deps):

The whole audio backend:
 * audio.cpp
 * audio_*.cpp
 * FmMidi: midisequencer.cpp, midisynth.cpp
 * decoder_*

Input backend:
 * input.cpp
 * input_*.cpp
 * keys.h

User interface:
 * baseui.h
 * *_ui

Basic Graphics API:
 * bitmap.cpp
 * color.cpp
 * drawable.cpp
 * font.cpp
 * graphics.cpp
 * hslrgb.cpp
 * image_*.cpp
 * sprite.cpp (maybe "Player", too specific)
 * plane.cpp
 * rect.cpp
 * text.cpp
 * tone.cpp
 * message_overlay.cpp
 * fps_overlay.cpp
 * pixel_format.h

Fonts:
 * bitmapfont_*.cpp
 * shinonome_*.cpp

Dependencies for this:
 * filefinder*.cpp
 * async_handler.cpp

This way fancy stuff like Steam integration or DRM-/NDA-API stuff can be added to engine core and all the stuff that wasted hundreds of dev hours is still under GPL in the Player.

Opinions?

**Comment by 20kdc** [comment]
Regarding basic graphics API, I'm curious as to why sprite.cpp is "maybe Player, too specific" - fitting with the Plane, Rect and Tone classes makes a very RGSS-style API, from which sprite.cpp would be something of an omission. (And on a side note, anything that could lead to the main screen composite being sufficiently abstract to get replaced *without* reimplementing bitmap.cpp is probably a good thing.)
Abstracting away Core might have other non-licensing-related benefits, specifically for anyone who particularly wants to make a reimplementation of Core with HW-acceleration, low as the chance is.

**Comment by Ghabry** [comment]
Didn't check all files when I wrote the previous comment. Rechecked sprite.cpp. You are right, besides the "BushDepth" function all of Sprite is very universal so it should be included in "Core".

**Comment by Ghabry** [comment]
Imo FileFinder can be removed from the dependency list because it is highly specific for RPG Maker 2000 features but we need a way to configure the main filesystem that is to be used... I added the WIP FileSystem API instead.

Player.cpp is needed for handling an update loop and some basic ui configuration but this must be refactored out in PlayerCore.cpp
Parts of Cache.cpp could be useful.

Made a quick compile with core dependencies only. Many files have useless includes but here are the criticial issues:

Problematic files that would need changes for a good core abstraction:
 * async_handler.cpp: Depends on FileFinder
 * audio.cpp: Depends on Player
 * audio_generic.cpp: Depends on FileFinder
 * bitmap: Depends on FileFinder
 * decoder_wildmidi.cpp: Depends on FileFinder
 * filesystem.cpp: Depends on FileFinder
 * font.cpp: Depends on Cache, FileFinder, Game_System, Player and ExFont --> Font configuration should be moved somewhere else
 * fps_overlay.cpp: Depends on Player (for speed modifier)
 * graphics.cpp: Minor dependency on "cache"
 * input.cpp: Depends on Player.cpp
 * message_overlay.cpp: Depends on Game_Message::WordWrap
 * output.cpp: Depends on FileFinder, Main_Data and Player
 * plane.cpp: Depends on Main_Data (for shake_position)
 * scene.cpp: Depends on Player::Update
 * sdl2_ui.cpp: Depends on Player::touch_flag/mouse_flag/exit_flag/no_audio_flag, Player::Pause, Player::Resume
 * text.cpp: Has exfont rendering, depends on Cache for the system graphic. Maybe should be replaced with a more generic renderer or use Font directly (will work when the font branch is in)

**Comment by Albeleon** [comment]
In regards to the GPL license exception that @Ghabry wrote two years ago, I also agree

**Comment by Ghabry** [comment]
There was some discussion about SteamShim. This is used by some games and said to not violate the GPL but this specific interprocess communication just to circumvent GPL is in my option a huge grey area (not sure if (64,64,64) or (192,192,192) though ;)) and wouldn't survive at a court.

For a short term solution, until all the other stuff is solved, just giving a general exception for SteamShim is maybe a good solution.

According to GPLv3 FAQ the exception is something like

```
If you modify this Program, or any covered work, by linking or combining it
with SteamShim (or a modified version of that library), containing parts covered
by the terms of MIT license, the licensors of this Program grant you additional
permission to convey the resulting work. Corresponding Source for a non-source
form of such a combination shall include the source code for the parts of SteamShim
library used as well as that of the covered work (excluding the Steamworks SDK).
```

**Comment by JeDaYoshi** [comment]
Anything regarding this nowadays?

**Comment by vegagames** [comment]
What is the current status of the license change?

**Comment by Ghabry** [comment]
It seems that more and more users are asking for it so the priority is raising but we made no significant progress yet :/.

**Comment by JeDaYoshi** [comment]
Yeah, just as someone interested on this - I know it should not be the highest priority, but... also that it's been requested for +6 years!

Now, what's stopping this from happening?

**Comment by Br4ssman** [comment]
We have a possible publisher interested to publish our next title.
This publisher only works with PS4, Vita, XBOX and Switch consoles, so changing the license will open EasyRPG to commercial projects and not only games for hobby.

Anyway, It's a bit dark for me how to publish on E-Shop or other consoles using EasyRPG.

**Comment by vegagames** [comment]
Having the license remain GPLv3 is a massive hindrance for people trying to make commercial games on consoles and other restrictive platforms. As much as I love the GPL, it's not fitting for this project. Please give us some kind of feedback as to what's going on. Even if the license change is no longer happening, it would be good to have some kind of closure or update on the issue.

**Comment by fmatthew5876** [comment]
I fully support changing the license to allow people to develop and sell commercial games. In order to help promote the engine and get more users, I suggest we try to resolve this sooner rather than later.

**Comment by Cloudef** [comment]
I'd like to sell a mobile app that would use easyrpg to launch rpg maker games, so I'd be very interested in a linking exception at least. I could workaround this by building easyrpg as separate APK and prompting for user to download / install it, whenever the feature would be needed. But having it already embedded into application as a shared library would be nice. Of course the source code for that library is mostly if not completely unchanged easyrpg, and any patches if needed I'd be ready to give upstream.

Another option would also be to provide a feature for the official APK to be able to launch a game from a `.zip` file called by another application which would also work for my use cases. (And any other user friendly features I would provide, such as prompting for download of RTP if missing and so on)

**Comment by Cloudef** [comment]
Never mind, I'm not violating GPL as easyrpg and mine are separate apps. If I do any modifications to easyrpg I have to host the source though.

**Comment by fmatthew5876** [comment]
I would like to follow up on this as I've now contributed a significant portion of the code in various parts of Player over the last 2 years.

I would like to use Player for a commercial game, but unfortunately linking and inheriting won't really be enough. I'll want to do a lot of things like modify `Game_Actor` directly and add more stats, or change behaviors in random other parts of the code that aren't really extensible.

Getting all the player interfaces I want to change to be cleanly extensible using inheritance or factories will take years more and really doesn't make sense.

What's the current thinking on license and what do we need to do to get there?

This would be a good candidate for next major release.

**Comment by Ghabry** [comment]
Is it problematic for you that you must publish your changes to ``Game_Actor`` and other classes under GPL?

The idea we talked above was that we need a solution for incompatible SDKs that require NDA. 
The proposed solution was a "core" library that provides the engine and the operating system abstraction. This one would be MIT license so you can keep it secret.

Then there is the "Player". This stays under GPL and talks to the "Core".
Player has a linking exception: You can link it against "Core" without publishing the "Core" code (=GPL does not apply) when the interface the Player uses for communication was not altered.

**Comment by Ghabry** [comment]
I'm getting older and this issue too. My preference by now is to just move to a liberal license such as MIT and call it done.

Imo this linking exception stuff just doesn't work. Besides this NDA stuff which can be somehow worked around there are many use cases which just doesn't work.

Lets say somebody wants to add some completely new feature to the Player (matthew always uses a battle system as an example, but there are more), then this is required per GPL to be also licensed under GPL. Even when this new scene is a complete novel invention and it only calls into Player data structures to e.g. get the Actor HP it is already a derived work. You could also start here to add stuff like "GPL exempt functions" like the Linux Kernel does but imo this does not really scale for such a small project (and is silly).

There is also a gray area here: Because event code is excluded from all of this one could just invent new event commands and do clever event scripting to bypass most GPL-problems. Here you have to raise the question then if this is a GPL-violation.

The GPL-conforming forks on github for games sold on Steam etc. are so low-quality commit wise. I cannot remember that we were ever able to backport anything from them.

There is also no real business model (anymore?) related to somebody taking our code and selling it: RPG Maker is a niche and they moved to Javascript. RPG Maker 2003 is abandoned.

Also the GPL does not protect against any type of "abuse" that happens since years: Taking open source code, filling it with Ads and putting it in the Android Store. Just look at "JoyPlay": They just took an Android-fork of mkxp, added some ruby-compat functions for Win32 and the app is full of ads and has a Patreon. This is totally against the purpose of these licenses imo but perfectly legal.

Asking for the current opinion @fdelapena and @carstene1ns on this.

-------

Btw for the editor I will not accept a license change ever. There GPLv3 is no big deal imo. I see no legitimate use case for closed-source editor releases and if you do not distribute your custom editor then there is no requirement to share the source anyway. (maybe one should even move to AGPL because many hostile cloud businesses bypass the GPL now via "X as a service")

**Comment by fmatthew5876** [comment]
I'm 100% in support for this. There is really no viable way to cleanly separate game specific customizations.

**Comment by fdelapena** [comment]
Nothing to add, I fully agree and liblcf is MIT already, so looks good to me :+1:, also would allow to move some Player code (e.g. map code) to liblcf without issues. 

---

Editor is a different thing, and in case of a potential plug-in mechanism with some visual interface and providing features for both Player and Editor, I don't see issues with a proper design.

---

Not just RPG Maker 2003 ended support since a while. Kadokawa (or now Gotcha Gotcha Games who owns the Tkool stuff in Japan) announced the last month they are [ending support for RPG Maker 2000 and all RGSS-based series, including VALUE! editions](https://rpgtkool.hatenablog.com/entry/2021/02/25/135545) on 2021-03-31. They won't reply for support requests, though they will continue to provide RTP downloads and selling them (shrug, I guess via third party vendors), but no further changes on them.

So it's expected third party vendors on other countries and languages (e.g. rpgmakerweb by Degica) to continue selling legacy products as is, but if something breaks, they won't do anything to fix it. There are also vendors with [support limitations for long time customers for old purchases to provide support due to law limitations](https://forums.rpgmakerweb.com/index.php?threads/due-to-international-privacy-laws-sales-support-ending-for-purchases-before-september-12th-2013.106241/). How they could even verify a customer has rights to distribute RTP assets? ;).

**Comment by Ghabry** [comment]
Two days ago I wrote about just using MIT because I'm sick of all this exception nonsense which is impossible to solve for non-lawyers.

Yesterday I read about the "Mozilla Public License 2.0". This one is also interesting and also solves the problems imo:

It is a copy-left license but it is a file-based copy-left. This means when you modify a file under MPL2.0 your code must be licensed under MPL2.0.

When you add new files they are not subject to the MPL2.0, so you can add new scenes, new drawables, new uis (for propreitary SDK stuff) under any license you want.

**Comment by fmatthew5876** [comment]
One use case I can think of is I want to change the set of stats from just ATK, DEF, SPI, AGI to add more. This and many other changes to `Game_Actor`.

Doing changes like that requires modifying code all over the place. While I could add a new child class to `Game_Battler`, the reality is that `Game_Actor` is used everywhere.


Also, is it worth the complexity of having difference licenses for different files? What are we trying to protect against here?

**Comment by Ghabry** [comment]
Well this was just a discussion point.
You could create a Game_ActorMatthew that inherits from Game_Actor and put your stuff in there.

Anyway. The only threat model for me (The MPL2 would not really help alot more here) is that some people maybe have the smart idea to create proprietary patches RPG_RT style that are only for the windows version. Then advertise this nonsense on RMN and people start creating games using this crap. Though everytime this happens we can just tell them that this sucks and then hopefully nobody uses them.

tl;dr I'm still fine with MIT

**Comment by Ghabry** [comment]
Sorry for pinging you again. It has been a few years and my last question was about adding a linking exception.
After thinking about this for years we do not think that this will work for a game engine where custom modifications are needed. Especially when the author wants to keep them propreitary for whatever reason (custom changes are not really useful for the upstream project). I just hope that some companies distributing a game bundled with the Player will still contribute back bug fixes or enhancements.

So the new question is (just answer with "I (dis)agree"):
_Do you agree with relicensing of the code to a non-copyleft-license such as the MIT license?_

@Zegeri
@scurest 
@ChristianBreitwieser
@Tondorian
@tyrone-sudeium 
@rohkea 
@rueter37

**Comment by scurest** [comment]
I agree. (Sup.)

**Comment by rueter37** [comment]
I agree.

**Comment by tyrone-sudeium** [comment]
I agree.

My preference would be the MIT license across the entire Player; MPL sounds nice but it also sounds like it adds a bunch of legal complexity and I'm not sure anybody here's _really_ passionate about legal complexity.

**Comment by rohkea** [comment]
I agree. (Also, my agreement is not needed, since all my contributions are already CC0.)

**Comment by Ghabry** [comment]
Oh, sorry. I didn't want to ping you. I confused you with the one who did the translation feature 😅 
@sorlok -> Please read my longer comment above https://github.com/EasyRPG/Player/issues/167#issuecomment-806572598

**Comment by ChristianBreitwieser** [comment]
I agree.

**Comment by sorlok** [comment]
>> Do you agree with relicensing of the code to a non-copyleft-license such as the MIT license?

I agree.

**Comment by carstene1ns** [comment]
I am also for relicensing to a permissive license such as ISC (or the equivalent MIT or BSD 2 clause) or Apache License version 2, *as long* as the new license is GPL compatible (which these are).
(+) As said there are static ports that have libraries under GPL, so linking them is still possible.
(+) We always get the free advertising with these (our copyright notice needs to be distributed somewhere within the binary or readme for example)
(-) However, we might not benefit of backports that much (which there are not many today)

**Comment by Ghabry** [comment]
Okay except 2 devs all have already answered, thanks! We have no real rush here so we can wait for the remaining one or two months more. Then we have to check if there code is still existant and if yes rewrite it, just to be safe.

I will open a follow up issue with problems we have to solve related to relicensing.

**Comment by Zegeri** [comment]
> Do you agree with relicensing of the code to a non-copyleft-license such as the MIT license?

I agree

**Comment by Ghabry** [comment]
I just had a quite long GPL-discussion with Ancurio and we actually found another worrying problem (which is easier to reach with Ruby than with the 2k-interpreter though):
https://www.gnu.org/licenses/gpl-faq.html#InterpreterIncompat

When you just drop a random game next to the Player it counts as "Data that is interpreted" which is fine.

But when you make a more tight integration, e.g. adding game specific features to Player such as highly specific new interpreter commands then it could be that the game counts as a "facility" making it a part of the Player --> GPL applies to the game.

This could be also a problem for Oneshot that uses a heavily modified mkxp.

Now I'm kinda happy to get rid of GPL, lots of gray area here. :thinking:

**Comment by florianessl** [comment]
Having stumbled upon this decade old issue after recently starting to contribute myself to this project, I'd like to clarify following:
I consider all my EasyRPG & liblcf related commits (both here and in any of my forks) up until now to be CC0 and considering my future contributions, relicensing under any of the spiffy licenses suggested here is also fine with me. 👍

**Comment by monokotech** [comment]
Please do not use the MIT license, as this will benefit proprietary software developers, but they will not be of much help to us.

From here, we can gain a better understanding of proprietary software:
https://www.gnu.org/proprietary/proprietary.html

**Comment by Mimigris** [comment]
The change of license that is being discussed is, for instance, to allow games using the engine to be sold on platforms where changes need to be made to the engine while having those changes not be public (e.g. game consoles), and in this case a license like MIT where the possibility to make changes private is required, which is one of the points of the discussion related to the licensing change. Do you have something else than MIT that meets those criterias that you would like to suggest?

**Comment by monokotech** [comment]
> The change of license that is being discussed is, for instance, to allow games using the engine to be sold on platforms where changes need to be made to the engine while having those changes not be public (e.g. game consoles), and in this case a license like MIT where the possibility to make changes private is required, which is one of the points of the discussion related to the licensing change. Do you have something else than MIT that meets those criterias that you would like to suggest?

In this case, they should choose another engine instead of EasyRPG.

**Comment by florianessl** [comment]
I myself don't have any issue with publishing any changes I'm making to the source code of EasyRPG and was planning to do so anyway. Even in case my own game project would be released commercially some day, I'd of course make all new additions available publicly, GPL or otherwise. I want others to profit from the planned new features/tools and make their own awesome games with them...
But regarding third-party libraries for consoles/Steam that require some "copyright" license, this might be problematic. If EasyRPG stayed on GPL for whatever reason, it would hopefully still be possible to negotiate some other license agreement with the devs?

**Comment by Rinnegatamante** [comment]
I think the problem with staying with GPL in light of possible license negotiation case-by-case is given by the fact every cotnributor should be part of said agreement decision and, given how lengthy and complex even just switching from GPL to MIT is being, I think it would be unrealistically handleable.

**Comment by monokotech** [comment]
The decline from GPL to MIT. You have forgotten why GPLv3 was originally used, it seems like you couldn't resist the temptation.

**Comment by monokotech** [comment]
I am just strongly advising against the change. I apologize if it makes you feel uncomfortable.

After EasyRPG Player changes its license, I should be able to relicense the latest changes/updates under the GPL license in my fork.

**Comment by Ghabry** [comment]
Sure, you can freely decide to stay on a different license. As your code changes are, when we add the callbacks, isolated in other files it will be easy to declare "The multiplayer folder/feature is GPL only".

**Comment by monokotech** [comment]
> Sure, you can freely decide to stay on a different license. As your code changes are, when we add the callbacks, isolated in other files it will be easy to declare "The multiplayer folder/feature is GPL only".

I am quite puzzled. If I make changes to source files in the `src` directory but want to keep my changes under the GPL, how should I handle merging, as this process seems to be getting more and more confusing.

**Comment by Rinnegatamante** [comment]
> Sure, you can freely decide to stay on a different license. As your code changes are, when we add the callbacks, isolated in other files it will be easy to declare "The multiplayer folder/feature is GPL only".

That's not possible. By the moment you include GPL code in your executable, the whole executable becomes GPL (since GPL is more restrictive). Either the submitted code must be relincensed to MIT or, if even technically possible, it could be dynamically loaded and published under LGPL rather than GPL.

Some reference: https://softwareengineering.stackexchange.com/questions/204410/how-are-gpl-compatible-licenses-like-mit-usable-in-gpl-programs-without-being-su

**Comment by rohkea** [comment]
> That's not possible. By the moment you include GPL code in your executable, the whole executable becomes GPL (since GPL is more restrictive).

I believe the plan was to have a flag like `--nonfree` when compiling, see #2500 

So, some files will exist in 2 variants, perhaps separated into different folders:

* MIT variant will only contain a stub: an empty function that does nothing, or something like that,
* GPL variant will contain some additional features.

Then the user will choose what variant they want:

* if they compile with `--nonfree` flag, they have MIT version (all code under MIT, nothing under GPL, but less features),
* if they compile with the default flags, they have GPL version (includes GPL code, so has more features).

**Comment by monokotech** [comment]
> That's not possible. By the moment you include GPL code in your executable, the whole executable becomes GPL (since GPL is more restrictive). Either the submitted code must be relincensed to MIT or, if even technically possible, it could be dynamically loaded and published under LGPL rather than GPL.
> 
> Some reference: https://softwareengineering.stackexchange.com/questions/204410/how-are-gpl-compatible-licenses-like-mit-usable-in-gpl-programs-without-being-su

I might not have been clear. What I meant was from MIT to GPL.

GPL + MIT: https://softwareengineering.stackexchange.com/questions/105912/can-you-change-code-distributed-under-the-mit-license-and-re-distribute-it-unde

https://softwarefreedom.org/resources/2007/gpl-non-gpl-collaboration.html

**Comment by Ghabry** [comment]
Hey. I'm pinging here all new contributors who haven't stated whether they agree to any license changing to a non-copyleft-license (such as MIT).

Please state in a comment whether you agree to relicense from GPLv3 to MIT license (or similar).

@Desdaemon 
@drxgb 
@enewey 
@MackValentine 
@MakoInfused 
@Primekick 
@ToolMan2k 
@tyrone-sudeium 

For difference of GPL vs. MIT see the loooong discussion here. The main difference is that code under MIT can have proprietary changes (so you are not required to publish them contrary to GPL). So the main difference is when _distributing_ the Player e.g. as part of a game and you have made custom changes to the game e.g. a new battle system.

Actively _hurting_ is GPL when publishing of games on platforms that are incompatible with GPL (such as console platforms) or linking against SteamWorks e.g. for Achievement support. So you cannot currently do this with our license.

There is actually a game planned to be ported to Nintendo Switch and they want to use EasyRPG Player. So would be good to have that roadblock out of the way ;).

_What happens if I do not agree?_: When it is an essential component the code will be removed. When it is something that is optional the code will be hidden behind a compiler flag (so they can opt-in whether the Player is subject under MIT or GPL).

---

btw from now on I will delete any heated discussion about why MIT will hurt the project.

**Comment by enewey** [comment]
Agreed!

**Comment by tyrone-sudeium** [comment]
I agree to relicensing any of my contributions from GPLv3 to a permissive license (MIT or similar).

**Comment by JorgeMaker8000** [comment]
Yes, I agree.

**Comment by MakoInfused** [comment]
@Ghabry I agree to the terms.

**Comment by johannschopplich** [comment]
@Ghabry I'm all in for the switch to MIT license!

**Comment by drxgb** [comment]
No problem, I agree. Let's goooo !! 🚀

**Comment by Primekick** [comment]
Not a problem for me, I agree to the relicense.

**Comment by MackValentine** [comment]
Okay for me

**Comment by BlisterB** [comment]
Ok for me

**Comment by 20kdc** [comment]
I agree that any and all of my EasyRPG contributions and work I have written for any fork of an EasyRPG project may be used under the MIT license.

**Comment by Ghabry** [comment]
Okay thanks everyone for agreeing. All the work required for it will be postboned for after 0.8.1

Don't want to delay the release even further 😉

**Comment by Ghabry** [comment]
I'm doing some more pinging here as I missed at least one relevant contributor:

- [ ] @MikuChan03 

----

The contributed code is mostly WQY which is GPL'ed but I'm still asking because why not:

- [ ] @DBLobster 

----

And while we are at it:

For the following users I don't think that the contribution has enough "Schöpfungshöhe" (Threshold of originality, the change is minor, only one/two lines) to be relevant for the licensing discussion but I'm still pinging you because I don't want that you feel that we consider your contribution irrelevant:

- [ ] @lychees 
- [ ] @lumiscosity 
- [ ] @elsemieni 
- [ ] @Shin-NiL

**Comment by MarianoGnu** [comment]
I will be very hapy to see this project changing to MIT lisence

**Comment by Shin-NiL** [comment]
MIT is the way to go

**Comment by elsemieni** [comment]
Of course, agree :)

**Comment by aidatorajiro** [comment]
yeah, I agree.

**Comment by DBLobster** [comment]
I agree.

**Comment by Ghabry** [comment]
@DBLobster  yeah we know that WQY is incompatible and cannot be used in a MIT build. That font will be disabled in that case.

This was just about your code excluding the WQY font.

**Comment by jetrotal** [comment]
@lychees also gave permission to move their code to MIT.

<img width="838" height="607" alt="Image" src="https://github.com/user-attachments/assets/f01f9aad-773a-49b9-a1fc-9d6a68c335be" />

**Comment by drxgb** [comment]
I agree. There's no problem to switch to MIT. I believe that it could be the better choice besides.
